src/irr/asset/COBJMeshFileLoader.cpp

```diff
@@ -39,13 +39,361 @@ namespace std
+#include <irr/builtin/glsl/brdf/specular/geom/ggx_smith.glsl>
+#include <irr/builtin/glsl/brdf/specular/fresnel/fresnel.glsl>
+
+void main()
+{
+    vec3 N = normalize(Normal);
+    //some approximation for computing tangents without UV
+    vec3 c1 = cross(N, vec3(0.0, 0.0, 1.0));
+    vec3 c2 = cross(N, vec3(0.0, 1.0, 0.0));
+    vec3 T = (dot(c1,c1) > dot(c2,c2)) ? c1 : c2;
+    T = normalize(T);
+    vec3 B = normalize(cross(N,T));
+    vec3 V = -ViewPos;
+
+    vec3 NdotV = dot(N,V);
+#define NdotL NdotV
+#define NdotH NdotV
+
+    vec3 color = PC.emissive*0.01;
+    if (NdotL > FLT_MIN)
+    {
+        float lightDistance2 = dot(V,V);
+        float Vrcplen = inversesqrt(lightDistance2);
+        NdotV *= Vrcplen;
+        V *= Vrcplen;
+
+        vec3 TdotV = dot(T,V);
+        vec3 BdotV = dot(B,V);
+#define TdotL TdotV
+#define BdotL BdotV
+#define TdotH TdotV
+#define BdotH BdotV
+
+        float at = sqrt(PC.roughness);
+        float ab = at*(1.0 - PC.anisotropy);
+
+        float fr = Fresnel_dielectric(PC.ior, NdotV);
+        float one_minus_fr = 1.0-fr;
+        float diffuseFactor = 1.0 - one_minus_fr*one_minus_fr;
+        float diffuse = 0.0;
+        if (PC.metallic < 1.0)
+        {
+            if (PC.roughness==0.0)
+                diffuse = 1.0/PI;
+            else
+                diffuse = oren_nayar(PC.roughness, N, V, V, NdotL, NdotV);
+        }
+        float specular = 0.0;
+        if (NdotV > FLT_MIN)
+        {
+            float ndf = GGXBurleyAnisotropic(PC.anisotropy, PC.roughness, TdotH, BdotH, NdotH);
+            float geom = GGXSmithHeightCorrelated_aniso_wo_numerator(at, ab, TdotL, TdotV, BdotL, BdotV, NdotL, NdotV);
+            specular = ndf*geom*fr;
+        }
+
+        color += (diffuseFactor*diffuse*PC.albedo + specular) * NdotL / lightDistance2;
+    }
+    OutColor = vec4(color*PC.transmissionFilter, 1.0);
+}
+)";
+    constexpr const char* VERT_SHADER_UV = 
+R"(#version 430 core
+
+layout (location = 0) in vec3 vPos;
+layout (location = 2) in vec2 vUV;
+layout (location = 3) in vec3 vNormal;
+
+layout (location = 0) out vec3 LocalPos;
+layout (location = 1) out vec3 ViewPos;
+layout (location = 2) out vec3 Normal;
+layout (location = 3) out vec2 UV;
+
+layout (set = 1, binding = 0, row_major, std140) uniform UBO {
+    mat4 MVP;
+    mat4 MV;
+    mat3 NormalMat;
+    vec3 EyePos;
+} CamData;
+
+void main()
+{
+    LocalPos = vPos;
+    gl_Position = CamData.MVP*vec4(vPos, 1.0);
+    ViewPos = (CamData.MV*vec4(vPos, 1.0)).xyz;
+    Normal = normalize(CamData.NormalMat*vNormal);
+    UV = vUV;
+}
+)";
+    constexpr const char* FRAG_SHADER_UV =
+R"(#version 430 core
+
+layout (location = 0) in vec3 LocalPos;
+layout (location = 1) in vec3 ViewPos;
+layout (location = 2) in vec3 Normal;
+layout (location = 3) in vec2 UV;
+layout (location = 0) out vec4 OutColor;
+
+#define ILLUM_MODEL_MASK 0x0fu
+#define map_Ka_MASK uint(1u<<4u)
```

Comment:
all these shaders should be simple overrides of the UV version (or vice versa), thats why I want you to wrap every blok in `#ifndef + #define` guards with separated declarations and definitions so simple changes like this are possible without duplicate code.

Reply:
actually make the non-UV version be the override of the UV version